src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl

```diff
@@ -0,0 +1,270 @@
+    const float box_h = y1 - y0 + COORDINATES_OFFSET;
+
+    proposals[proposal_idx + 0] = x0;
+    proposals[proposal_idx + 1] = y0;
+    proposals[proposal_idx + 2] = x1;
+    proposals[proposal_idx + 3] = y1;
+    proposals[proposal_idx + 4] = ((MIN_SIZE <= box_w) && (MIN_SIZE <= box_h)) ? score : 0.f;
+}
+
+#    undef COORDINATES_OFFSET
+
+#endif /* EDGPSI_STAGE_0 */
+
+#ifdef EDGPSI_STAGE_1
+
+typedef struct __attribute__((__packed__)) {
+    float x0;
+    float y0;
+    float x1;
+    float y1;
+    float score;
+} Box;
+
+inline void FUNC(swap_box)(__global Box* a, __global Box* b) {
+    const Box temp = *a;
+    *a = *b;
+    *b = temp;
+}
+
+inline int FUNC(partition)(__global Box* arr, int l, int h) {
+    INPUT0_TYPE pivotScore = arr[h].score;
+    int i = (l - 1);
+    for (int j = l; j <= h - 1; j++) {
+        if (arr[j].score > pivotScore) {
+            i++;
+            FUNC_CALL(swap_box)(&arr[i], &arr[j]);
+        }
+    }
+    FUNC_CALL(swap_box)(&arr[i + 1], &arr[h]);
+    return (i + 1);
+}
+
+inline void FUNC(bubbleSortIterative)(__global Box* arr, int l, int h) {
+    for (int i = 0; i < h - l; i++) {
+        bool swapped = false;
+        for (int j = l; j < h - i; j++) {
+            if ((arr[j].score > arr[j + 1].score)) {
+                FUNC_CALL(swap_box)(&arr[j], &arr[j + 1]);
+                swapped = true;
+            }
+        }
+
+        if (!swapped)
+            break;
+    }
+}
+
+inline void FUNC(quickSortIterative)(__global Box* arr, int l, int h) {
+    // Create an auxiliary stack
+    const int kStackSize = 100;
+    int stack[kStackSize];
+
+    // initialize top of stack
+    int top = -1;
+
+    // push initial values of l and h to stack
+    stack[++top] = l;
+    stack[++top] = h;
+
+    // Keep popping from stack while is not empty
+    while (top >= 0) {
+        // Pop h and l
+        h = stack[top--];
+        l = stack[top--];
+
+        // Set pivot element at its correct position
+        // in sorted array
+        int p = FUNC_CALL(partition)(arr, l, h);
+
+        // If there are elements on left side of pivot,
+        // then push left side to stack
+        if (p - 1 > l) {
+            if (top >= (kStackSize - 1)) {
+                FUNC_CALL(bubbleSortIterative)(arr, l, p - 1);
+            } else {
+                stack[++top] = l;
+                stack[++top] = p - 1;
+            }
+        }
+
+        // If there are elements on right side of pivot,
+        // then push right side to stack
+        if (p + 1 < h) {
+            if (top >= (kStackSize - 1)) {
+                FUNC_CALL(bubbleSortIterative)(arr, p + 1, h);
+            } else {
+                stack[++top] = p + 1;
+                stack[++top] = h;
+            }
+        }
+    }
+}
+
+// 1. Sort boxes by scores
+KERNEL(edgpsi_ref_stage_1)(__global INPUT0_TYPE* proposals) {
+    __global Box* boxes = (__global Box*)proposals;
+
+    FUNC_CALL(quickSortIterative)(boxes, 0, NUM_PROPOSALS-1);
```

Comment:
if `pre_nms_count` is expected to be significantly smaller than total number of proposals, using partial sort here could improve performance quite a bit...

Reply:
@tgubanova-lohika Is there any on-going work to resolve this?

Reply:
not for now, i propose to leave in ref impl as it is

Reply:
we can rework it when will use common sorting for GPU